### ISSUE TITLE
install-app-deps hangs on a yarn workspace where all dependencies are hoisted to the root workspace

### ISSUE THREAD

**Issue by mariano-filipe** [electron-userland/electron-builder#6448]
* **Electron-Builder Version**: 22.14.5

* **Node Version**: 16.11.1

* **Electron Version**: 16.0.2
* **Electron Type (current, beta, nightly)**: current

* **Target**: win:nsis x64 and ia32

Upon upgrading the dependencies of a monorepo I maintain, I ended up with a situation where I had one workspace (say package1) with all its installed dependencies hoisted to the root workspace. This turned out to be a problem for the `electron-builder install-app-deps` script which hangs spawning multiple `yarn install` threads until my machine simply hangs without enough RAM.

The situation of my monorepo is something like this:
- root
  - node_modules // root deps and other hoisted deps from children packages
  - packages
    - package1
      - . // no node_modules, all deps were hoisted to the root node_modules
    - package2
      - node_modules
    - package3
      - node_modules

I did some debugging on the `install-app-deps` script and found out that it's not correctly detecting that package1 in the example above is already installed. The script simply checks if a `node_modules` folder or a `.pnp.js` file exists to determine if the package is installed, but as I explained above in some situations yarn may decide to hoist all deps to the root node_modules. I think the problematic function is this one:
https://github.com/electron-userland/electron-builder/blob/e95afc1ab8c1a09fc8c9496084fc9f49b185469e/packages/app-builder-lib/src/util/yarn.ts#L10-L31

This function keeps spawning new install threads because it never succeeds in detecting that package1 is already installed.

**Comment by ofsaleem** [comment]
yeah, i think im having the same problem. it hangs forever and eats all my resources
![image](https://user-images.githubusercontent.com/7631554/146853581-266365c1-8d75-4f92-96c4-d669493e8f9a.png)

**Comment by julianhille** [comment]
Having the same issue after:
- upgrade yarn 2.4.3 ro 3.2
- update node from 14.17 to 16
- electron 13 to 16

Actually a symlink from workspace/node_modules dir to the root node_modules dir works.

**Comment by julianhille** [comment]
Ok found out why it exactly started to break. my workspace package had a module version conflict with another package and so the `node_modules` dir existed.

**Comment by ian-oneleet** [comment]
This happened in an `npm` workspace where some electron dependencies were mistakenly deleted from package-lock.json.

The `"postinstall": "electron-builder install-app-deps"` script ran `npm install`, which didn't install what electron-builder was looking for but did run `postinstall`, which ran `npm install`, which didn't install what electron-builder was looking for but did run `postinstall`, which …

**Comment by ian-oneleet** [comment]
Still an issue. This specifically happens with both `npm` and `yarn` workspaces, which hoist dependencies to the root workspace. `electron-builder` endlessly tries to install a dependency, and never clues in that it's already installed. It ends up crashing the machine.

**Comment by beyondkmp** [comment]
@ian-oneleet  Could you help provide a minimum reproduced demo?

**Comment by ian-oneleet** [comment]
@beyondkmp Please clone [this repo](https://github.com/ian-oneleet/electron-builder-infinite-recursion-bug) and run `npm i`.

At least on my macOS system, this hangs forever. If you check Activity Monitor, you'll see a growing list of processes named "npm install". If you wait too long, it will exhaust system memory.

Edit: and thanks for looking into this!
